web/backend-fastapi/app/models/rag.py

```diff
@@ -65,29 +74,52 @@ def create_prompt(self, query: str, context_str: str, chat_history: List[ChatHis
     def get_response(self, query: str) -> str:
         bedrock_response = bedrock.get_response(query)
         return bedrock_response
-    
+
     def formatoutput(self, topk, lm_output):
         prettier = {}
-        prettier['llm'] = lm_output
-        prettier['topk'] = []
+        prettier["llm"] = lm_output
+        prettier["topk"] = []
         for k in topk:
             k_obj = {}
-            k_obj['score'] = k['score']
-            k_obj['ActId'] = k['node.ActId']
-            k_obj['Regulations'] = k['Regulations']
-            k_obj['sectionId'] = k['node.sectionId']
-            k_obj['sectionName'] = k['node.sectionName']
-            k_obj['url'] = k['node.url']
-            k_obj['text'] = k['text']
-            if 'references' in k:
-                k_obj['references'] = k['references']
-            prettier['topk'].append(k_obj)
+            k_obj["score"] = k["score"]
+            k_obj["ActId"] = k["node.ActId"]
+            k_obj["Regulations"] = k["Regulations"]
+            k_obj["sectionId"] = k["node.sectionId"]
+            k_obj["sectionName"] = k["node.sectionName"]
+            k_obj["url"] = k["node.url"]
+            k_obj["text"] = k["text"]
+            if "references" in k:
+                k_obj["references"] = k["references"]
+            prettier["topk"].append(k_obj)
         return prettier
 
+    @instrument
+    def re_rank_reference(self, topk, compared_text, doc_fields=["text"]):
+        model = CrossEncoder("cross-encoder/ms-marco-MiniLM-L-6-v2")
+        for field_index, field in enumerate(doc_fields):
+            # Map topk to [compared_text, doc_field] pairs list
+            pairs = [[compared_text, doc[field]] for doc in topk]
+            # Produces list of float values. Higher is more closely related. Should be parallel with pairs list.
+            scores = model.predict(pairs)
+            # Apply these scores to the original objects
+            for index, _ in enumerate(topk):
+                topk[index][f"cross_score_{field_index}"] = scores[index]
+        # Sort by these new scores
+        topk.sort(key=lambda x: tuple(x[field] for field in doc_fields), reverse=True)
+        return topk
+
     @instrument
     def query(self, query: str, chat_history: List[ChatHistory], embeddings, kg) -> str:
         context_str = self.retrieve(query, embeddings, kg)
+        # Rerank once to order for LLM
+        context_str = self.re_rank_reference(context_str, query)
         create_prompt = self.create_prompt(query, context_str, chat_history)
         bedrock_response = self.get_response(create_prompt)
+        # Rerank again to sort references by relevance to response
+        context_str = self.re_rank_reference(
```

Comment:
Why do we need a double rerank? Is the performance better with dobule rerank? seems strange.

Reply:
I would be good to remove the initial rerank prior to the bedrock request. I'm not convinced that it improves bedrock's response, but that would mean not reranking the references according to the user's query, as suggested in standup.

If you mean why rerank on two different fields, it seemed more consistent to prioritize which Act the information came from then order the references based on their text relevance. Just doing it by text didn't always land the expected Act at the top of the list.